### PR TITLE
RWA-1707 handle 500 errors

### PR DESCRIPTION
**Pull request by serdaroguzhanakin**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/RWA-1707

### Change description ###

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[X] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/controllers/advice/ApplicationProblemControllerAdvice.java [hunk 1]
```diff
@@ -58,10 +60,16 @@
...
 
-    @ExceptionHandler(FeignException.ServiceUnavailable.class)
+    @ExceptionHandler({
+        FeignException.ServiceUnavailable.class,
+        FeignException.FeignServerException.class,
+        FeignException.GatewayTimeout.class,
+        FeignException.BadRequest.class,
+        ServerErrorException.class,
+    })
     public ResponseEntity<ThrowableProblem> handleFeignServiceUnavailableException(FeignException ex) {
```
Comment (raghera) [inline src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/controllers/advice/ApplicationProblemControllerAdvice.java:73]: Should probably rename this method now as it deals with more exceptions than before.

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/controllers/advice/ApplicationProblemControllerAdvice.java [hunk 2]
```diff
@@ -58,10 +60,16 @@
 })
 @RequestMapping(produces = APPLICATION_PROBLEM_JSON_VALUE, consumes = APPLICATION_JSON_VALUE)
 @SuppressWarnings({"PMD.ExcessiveImports", "PMD.DataflowAnomalyAnalysis",
-    "PMD.UseStringBufferForStringAppends", "PMD.LawOfDemeter"})
+    "PMD.UseStringBufferForStringAppends", "PMD.LawOfDemeter", "PMD.CouplingBetweenObjects"})
 public class ApplicationProblemControllerAdvice extends BaseControllerAdvice implements ValidationAdviceTrait {
 
-    @ExceptionHandler(FeignException.ServiceUnavailable.class)
+    @ExceptionHandler({
+        FeignException.ServiceUnavailable.class,
```
Comment (raghera) [inline src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/controllers/advice/ApplicationProblemControllerAdvice.java:67]: which circumstances is this one thrown?

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/exceptions/v2/TaskAlreadyClaimedException.java [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+@SuppressWarnings("java:S110")
+public class TaskAlreadyClaimedException extends AbstractThrowableProblem {
+
+    private static final long serialVersionUID = -7314118612753159016L;
+
+    private static final URI TYPE = URI.create("https://github.com/hmcts/wa-task-management-api/problem/task-already-claimed-error");
+    private static final String TITLE = "Task Already Claimed Error";
+    
+    public TaskAlreadyClaimedException(String message, Exception ex) {
```
Comment (raghera) [inline src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/exceptions/v2/TaskAlreadyClaimedException.java:18]: Should add this new problem to the docs.  Here: https://tools.hmcts.net/confluence/pages/viewpage.action?pageId=1482336467&__ncforminfo=hnDfZ75336XaURs1yHOf7u7qiVTFAMDO8FVkeG4nMeYCuriPfu7BGG2UwvR2GlqLM0ukwbv0dRXr4Fp_SIuFdROVn6VChbl7

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/services/CamundaService.java [hunk 4]
```diff
@@ -429,10 +429,11 @@ private Map<String, CamundaVariable> performGetVariablesAction(String id) {
         try {
             return camundaServiceApi.getVariables(authTokenGenerator.generate(), id);
         } catch (FeignException ex) {
+            log.error("There was a problem fetching the variables for task with id: {} \n{}", id, ex.getMessage());
             throw new ResourceNotFoundException(String.format(
                 "There was a problem fetching the variables for task with id: %s",
                 id
-            ), ex);
```
Comment (raghera) [inline src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/services/CamundaService.java:435]: Don't you want to keep the original exception details?  Also you are catching a very generic exception here so how do we know that it was a 404?  Could possibly be a 5xx / 4xx error?

Error log is prob ok but ResourceNotFoundException assumes a 404?

File: src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/wataskconfigurationapi/controllers/TaskConfigurationControllerTest.java [hunk 5]
```diff
@@ -362,17 +365,45 @@ void should_succeed_and_return_configuration_with_new_role_name_case_manager() t
...
+        doThrow(FeignException.BadRequest.class)
+            .when(ccdDataServiceApi)
+            .getCase(any(), any(), any());
+
+        mockMvc.perform(
+                post(TASK_CONFIGURATION_ENDPOINT + testTaskId)
+                    .contentType(APPLICATION_JSON_VALUE)
+            )
+            .andExpectAll(
+                status().is5xxServerError(),
```
Comment (raghera) [inline src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/wataskconfigurationapi/controllers/TaskConfigurationControllerTest.java:384]: Should this not check that it is a 502?

File: src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/controllers/PostTaskForSearchCompletionControllerTest.java [hunk 6]
```diff
@@ -376,6 +381,56 @@ void should_return_a_200_with_empty_list_when_the_user_did_not_have_any_roles()
...
+
+        mockMvc.perform(
+                post("/task/search-for-completable")
+                    .header(AUTHORIZATION, IDAM_AUTHORIZATION_TOKEN)
+                    .header(SERVICE_AUTHORIZATION, SERVICE_AUTHORIZATION_TOKEN)
+                    .content(asJsonString(searchEventAndCase))
+                    .contentType(MediaType.APPLICATION_JSON_VALUE)
+            )
+            .andExpectAll(
+                status().is5xxServerError(),
```
Comment (raghera) [inline src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/controllers/PostTaskForSearchCompletionControllerTest.java:419]: should check the status is 502?

**Review by raghera — DISMISSED** [review]

**Review by raghera — COMMENTED** [review]
See comments

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/controllers/advice/ApplicationProblemControllerAdvice.java (hunk 1 above)
Comment (serdaroguzhanakin) [reply src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/controllers/advice/ApplicationProblemControllerAdvice.java:73]: renamed

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/exceptions/v2/TaskAlreadyClaimedException.java (hunk 3 above)
Comment (serdaroguzhanakin) [reply src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/exceptions/v2/TaskAlreadyClaimedException.java:18]: line 24 added to related document, please check

File: src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/wataskconfigurationapi/controllers/TaskConfigurationControllerTest.java (hunk 5 above)
Comment (serdaroguzhanakin) [reply src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/wataskconfigurationapi/controllers/TaskConfigurationControllerTest.java:384]: This method checks error code starts with 5 or not.
Here is their explanation 'Assert the response status code is in the 5xx range'
Should I take any action?

File: src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/controllers/PostTaskForSearchCompletionControllerTest.java (hunk 6 above)
Comment (serdaroguzhanakin) [reply src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/controllers/PostTaskForSearchCompletionControllerTest.java:419]: This method checks error code starts with 5 or not.
Here is their explanation 'Assert the response status code is in the 5xx range'
Should I take any action?

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/controllers/advice/ApplicationProblemControllerAdvice.java (hunk 2 above)
Comment (serdaroguzhanakin) [reply src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/controllers/advice/ApplicationProblemControllerAdvice.java:67]: here is an example ccd prod log 
[logs](https://portal.azure.com/#blade/AppInsightsExtension/BladeRedirect/BladeName/searchV1/ResourceId/%252Fsubscriptions%252F8999dec3-0104-4a27-94ee-6588559729d1%252FresourceGroups%252Fccd-shared-prod%252Fproviders%252FMicrosoft.Insights%252Fcomponents%252Fccd-prod/BladeInputs/%7B%22tables%22%3A%5B%22availabilityResults%22%2C%22requests%22%2C%22exceptions%22%2C%22pageViews%22%2C%22traces%22%2C%22customEvents%22%2C%22dependencies%22%5D%2C%22timeContextWhereClause%22%3A%22%7C%20where%20timestamp%20%3E%20datetime(%5C%222022-09-12T16%3A08%3A52.000Z%5C%22)%20and%20timestamp%20%3C%20datetime(%5C%222022-09-12T16%3A08%3A54.000Z%5C%22)%22%2C%22filterWhereClause%22%3A%22%7C%20where%20*%20has%20%5C%22ServiceUnavailable%5C%22%7C%20order%20by%20timestamp%20desc%22%2C%22originalParams%22%3A%7B%22eventTypes%22%3A%5B%7B%22value%22%3A%22availabilityResult%22%2C%22tableName%22%3A%22availabilityResults%22%2C%22label%22%3A%22Availability%22%7D%2C%7B%22value%22%3A%22request%22%2C%22tableName%22%3A%22requests%22%2C%22label%22%3A%22Request%22%7D%2C%7B%22value%22%3A%22exception%22%2C%22tableName%22%3A%22exceptions%22%2C%22label%22%3A%22Exception%22%7D%2C%7B%22value%22%3A%22pageView%22%2C%22tableName%22%3A%22pageViews%22%2C%22label%22%3A%22Page%20View%22%7D%2C%7B%22value%22%3A%22trace%22%2C%22tableName%22%3A%22traces%22%2C%22label%22%3A%22Trace%22%7D%2C%7B%22value%22%3A%22customEvent%22%2C%22tableName%22%3A%22customEvents%22%2C%22label%22%3A%22Custom%20Event%22%7D%2C%7B%22value%22%3A%22dependency%22%2C%22tableName%22%3A%22dependencies%22%2C%22label%22%3A%22Dependency%22%7D%5D%2C%22timeContext%22%3A%7B%22durationMs%22%3A2000%2C%22endTime%22%3A%222022-09-12T16%3A08%3A54.000Z%22%7D%2C%22filter%22%3A%5B%5D%2C%22searchPhrase%22%3A%7B%22originalPhrase%22%3A%22ServiceUnavailable%22%2C%22_tokens%22%3A%5B%7B%22conjunction%22%3A%22and%22%2C%22value%22%3A%22ServiceUnavailable%22%2C%22isNot%22%3Afalse%2C%22kql%22%3A%22%20*%20has%20%5C%22ServiceUnavailable%5C%22%22%7D%5D%7D%2C%22sort%22%3A%22desc%22%7D%7D)

File: src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/wataskconfigurationapi/controllers/TaskConfigurationControllerTest.java (hunk 5 above)
Comment (serdaroguzhanakin) [reply src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/wataskconfigurationapi/controllers/TaskConfigurationControllerTest.java:384]: This method checks error code starts with 5 or not.
Here is their explanation 'Assert the response status code is in the 5xx range'
Should I take any action?

Comment (raghera) [reply src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/wataskconfigurationapi/controllers/TaskConfigurationControllerTest.java:384]: My point is that you are checking that it is any 5xx error when it fact your test should be checking it is specifically a 502 status code?  
If someone later changes the status to 503 or 500 then this test would still pass but it's not correct.

File: src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/controllers/PostTaskForSearchCompletionControllerTest.java (hunk 6 above)
Comment (raghera) [reply src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/controllers/PostTaskForSearchCompletionControllerTest.java:419]: as previous comment

Comment (serdaroguzhanakin) [reply src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/controllers/PostTaskForSearchCompletionControllerTest.java:419]: in line 424 also check it 
jsonPath("$.status").value(502),

Comment (raghera) [reply src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/controllers/PostTaskForSearchCompletionControllerTest.java:419]: No this is different.  That's in the body more for debugging convenience.  The overall http status is not in the body.

**Review by raghera — DISMISSED** [review]

**Review by mkayad — APPROVED** [review]

**Pull request merged**